x/mongo/driver/topology/rtt_monitor.go

```diff
@@ -39,7 +39,12 @@ type rttConfig struct {
 }
 
 type rttMonitor struct {
-	mu            sync.RWMutex // mu guards samples, offset, minRTT, averageRTT, and averageRTTSet
+	mu sync.RWMutex // mu guards samples, offset, minRTT, averageRTT, and averageRTTSet
+
+	// connMu guards connecting and disconnecting. This is necessary since
+	// disconnecting will await the cancelation of a started connection. The
+	// use case for rttMonitor.connect needs to be goroutine safe.
+	connMu        sync.Mutex
```

Comment:
This PR requires the the rttMonitor's connection be established in a separate thread than the one a server.Disconnect() is called in:
![concurrency (2)](https://github.com/mongodb/mongo-go-driver/assets/24281431/1dccc8f7-89d7-44e8-863a-4fa68630261b)

Because of this, it's possible that a disconnect could occur before a connection actually takes place (P_0). This would make any call to the shared wait group unsafe. `connMu` guards this case.

Reply:
@qingyang-hu @matthewdale I just caught a race condition in the logic. Hopefully this is the last review required